### PR TITLE
Prepare for lockfile creation

### PR DESCRIPTION
**Pull request by msfjarvis**
Set up serde and a struct that we will be persisting into our lockfile

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -21,14 +29,19 @@ fn main() -> anyhow::Result<()> {
...
-                    entry.path().to_string_lossy().to_string(),
-                    repo.head()?.name().unwrap_or("None")
-                ));
+                let head = repo.head()?;
+                if let Some(head) = head.name() {
+                    repos.push(PersistableRepo {
+                        // Ideally we wanna do this, but it moves `dir`.
+                        // path: entry.path().to_string_lossy().strip_prefix(dir).unwrap().to_string(),
+                        path: entry.path().to_string_lossy().to_string(),
+                        remote_url: repo.remotes()?.get(0).unwrap_or("None").to_string(),
```
Comment (SphericalKat) [inline src/main.rs:38]: Is this the optimal remote to use?

Comment (msfjarvis) [reply src/main.rs:38]: Nope, we ideally want to use the upstream remote for the currently checked out branch. I haven't figured out how to do that yet.

File: src/main.rs [hunk 2]
```diff
@@ -21,14 +29,31 @@ fn main() -> anyhow::Result<()> {
...
-                ));
+                let head = repo.head()?;
+                if let Some(head) = head.name() {
+                    if let Ok(upstream) = repo.branch_upstream_name(head) {
+                        if let Ok(remote) = repo.find_remote(
+                            // This is a rather ugly hack, but not sure how else to get the required name
+                            // doesn't seem to work with the full name such as `refs/remotes/origin/master`
+                            upstream
+                                .as_str()
+                                .unwrap_or("None")
```
Comment (msfjarvis) [inline src/main.rs:40]: This should just become an `.unwrap()` so we don't get an unhelpful OOB panic later down this chain of methods.

Comment (SphericalKat) [reply src/main.rs:42]: Right, making the changes

File: gitice.lock [hunk 3]
```diff
@@ -0,0 +1,19 @@
+["../gitice"]
```
Comment (msfjarvis) [inline gitice.lock:1]: Might wanna add `gitice.lock` to gitignore.

File: src/main.rs [hunk 4]
```diff
@@ -10,25 +18,44 @@ fn main() -> anyhow::Result<()> {
...
+                            // This is a rather ugly hack, but not sure how else to get the required name
+                            // doesn't seem to work with the full name such as `refs/remotes/origin/master`
+                            upstream.as_str().unwrap().split('/').collect::<Vec<&str>>()[2],
+                        ) {
+                            let path = entry.path().to_string_lossy().to_string();
+                            repos.insert(
+                                path.clone(),
+                                PersistableRepo {
+                                    // Ideally we wanna do this, but it moves `dir`.
+                                    // path: entry.path().to_string_lossy().strip_prefix(dir).unwrap().to_string(),
```
Comment (msfjarvis) [inline src/main.rs:47]: Really need to figure out a way of making this work.

**Pull request merged**